### DISCUSSION TITLE
In-process code coverage collection

### DISCUSSION THREAD

**Discussion by mindplay-dk** [coverlet-coverage/coverlet#1359] [Q&A]
Hello,

I am writing a very small test-framework for C#, mostly as a fun exercise. 🙂

I wanted to add code coverage, somehow - am I correct to think that this is not something the platform SDK itself provides, and this library is essentially the best/only/closest thing to an "official" option for that?

Would the [`CoverletInProcDataCollector`](https://github.com/coverlet-coverage/coverlet/blob/master/src/coverlet.collector/InProcDataCollection/CoverletInProcDataCollector.cs) be the simplest/most appropriate facility to achieve that?

There is no inline documentation, and I don't see an example, nor any usages in any code I could find online.

How would I begin to use this?

Thanks!

**Comment by MarcoRossignoli** [comment]
Depends on how you want to provide the code coverage, you have 2 way static instrumentation(write the assemblies before to run) and dynamic instrumentation(use a .NET native profiler and instrument IL on the fly).
Coverlet doesn't expose the "instrumentation engine" as library and so if you want to use it you have to fork and expose what you need.
A good starting point is debug this test https://github.com/coverlet-coverage/coverlet/blob/master/test/coverlet.core.tests/Coverage/CoverageTests.SelectionStatements.cs#L15 
You should update the line 23 with `RunInProcess` so you can debug the instrumentation and understand how things works.

Another way is to start your test under a coverage and for that you can use the coverlet global tool(for static instrumentation) https://github.com/coverlet-coverage/coverlet#net-global-tool-guide-suffers-from-possible-known-issue or the MS provided one(dynamic) https://learn.microsoft.com/en-us/dotnet/core/additional-tools/dotnet-coverage

**Comment by Bertk** [comment]
Closing this discussion since it’s no longer relevant to the current state of the project.
